### ISSUE TITLE
Azure Pipelines Agent fails to run a job when release pipeline with many tasks is triggered

### ISSUE THREAD

**Issue by sponte** [microsoft/azure-pipelines-agent#2637]
## Have you tried troubleshooting?
Yes

## Agent Version and Platform
```
*****************************************************************************************
[2019-11-28 18:07:55Z WARN HostContext] **                                                                                     **
[2019-11-28 18:07:55Z WARN HostContext] ** Http trace is enabled, all your http traffic will be dumped into agent diag log.    **
[2019-11-28 18:07:55Z WARN HostContext] ** DO NOT share the log in public place! The trace may contains secrets in plain text. **
[2019-11-28 18:07:55Z WARN HostContext] **                                                                                     **
[2019-11-28 18:07:55Z WARN HostContext] *****************************************************************************************
[2019-11-28 18:07:55Z INFO AgentProcess] Agent built on Linux (X64) - linux-x64.
[2019-11-28 18:07:55Z INFO AgentProcess] Agent built for Linux (X64).
[2019-11-28 18:07:55Z INFO AgentProcess] Running on Linux (X64).
[2019-11-28 18:07:55Z INFO AgentProcess] RuntimeInformation: Linux 3.10.0-693.58.1.el7.x86_64 #1 SMP Tue Aug 13 11:39:02 UTC 2019.
[2019-11-28 18:07:55Z INFO AgentProcess] Version: 2.160.1
[2019-11-28 18:07:55Z INFO AgentProcess] Commit: 8c04bd1029f0b89ede934e870f0eac413da3156c
[2019-11-28 18:07:55Z INFO AgentProcess] Culture: en-US
[2019-11-28 18:07:55Z INFO AgentProcess] UI Culture: en-US
```

## Azure DevOps Type and Version
dev.azure.com
https://xlcatlin.visualstudio.com

## What's not working?
Release pipelines with large enough number of tasks fail to start up giving following error message. Mind you, most of the time the location is around 32k:

```
##[Error 1]
[2019-11-28 18:09:43Z WARN HostContext] *****************************************************************************************
[2019-11-28 18:09:43Z WARN HostContext] **                                                                                     **
[2019-11-28 18:09:43Z WARN HostContext] ** Http trace is enabled, all your http traffic will be dumped into agent diag log.    **
[2019-11-28 18:09:43Z WARN HostContext] ** DO NOT share the log in public place! The trace may contains secrets in plain text. **
[2019-11-28 18:09:43Z WARN HostContext] **                                                                                     **
[2019-11-28 18:09:43Z WARN HostContext] *****************************************************************************************
[2019-11-28 18:09:43Z INFO Program] Version: 2.160.1
[2019-11-28 18:09:43Z INFO Program] Commit: 8c04bd1029f0b89ede934e870f0eac413da3156c
[2019-11-28 18:09:43Z INFO Program] Culture: en-US
[2019-11-28 18:09:43Z INFO Program] UI Culture: en-US
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Bin': '/vsts/agent/bin'
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Root': '/vsts/agent'
[2019-11-28 18:09:44Z INFO HostContext] Well known config file 'Proxy': '/vsts/agent/.proxy'
[2019-11-28 18:09:44Z INFO VstsAgentWebProxy] Config proxy at: http://10.64.80.180:8080.
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Bin': '/vsts/agent/bin'
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Root': '/vsts/agent'
[2019-11-28 18:09:44Z INFO HostContext] Well known config file 'ProxyCredentials': '/vsts/agent/.proxycredentials'
[2019-11-28 18:09:44Z INFO VstsAgentWebProxy] Config proxy use DefaultNetworkCredentials.
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Bin': '/vsts/agent/bin'
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Root': '/vsts/agent'
[2019-11-28 18:09:44Z INFO HostContext] Well known config file 'ProxyBypass': '/vsts/agent/.proxybypass'
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Bin': '/vsts/agent/bin'
[2019-11-28 18:09:44Z INFO HostContext] Well known directory 'Root': '/vsts/agent'
[2019-11-28 18:09:44Z INFO HostContext] Well known config file 'Certificates': '/vsts/agent/.certificates'
[2019-11-28 18:09:44Z INFO AgentCertificateManager] No certificate setting found.
[2019-11-28 18:09:44Z INFO Worker] Waiting to receive the job message from the channel.
[2019-11-28 18:09:44Z INFO Worker] Message received.
[2019-11-28 18:09:45Z ERR  Program] Newtonsoft.Json.JsonReaderException: After parsing a value an unexpected character was encountered: m. Path 'plan.artifactUri', line 1, position 30733.
   at Newtonsoft.Json.JsonTextReader.ParsePostValue(Boolean ignoreComments)
   at Newtonsoft.Json.JsonTextReader.Read()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize[T](JsonReader reader)
   at Microsoft.VisualStudio.Services.WebApi.JsonUtility.FromString[T](String toDeserialize, JsonSerializerSettings settings)
   at Microsoft.VisualStudio.Services.Agent.Worker.Worker.RunAsync(String pipeIn, String pipeOut)
   at Microsoft.VisualStudio.Services.Agent.Worker.Program.MainAsync(IHostContext context, String[] args)
[2019-11-28 18:09:45Z INFO HttpTrace] DiagListeners finished transmitting data.

```

I have a lot more logging information but it agent was ran with trace enabled, I won't be sharing these publicly.

Please find example release pipeline attached.
[POC-Xray-CD.zip](https://github.com/microsoft/azure-pipelines-agent/files/3902811/POC-Xray-CD.zip)

**Comment by sponte** [comment]
Hello, any updates on this issue? Do you need any more information?

**Comment by damccorm** [comment]
Hey @sponte sorry for the slow response - it sounds like maybe our JsonReader is falling over because the file is getting too big?

Could you share a little bit more about your use case here? Why do you need 32k tasks? It also doesn't look like the release pipeline you attached has that many tasks, I only see 4

**Comment by sponte** [comment]
Hello, thanks for getting back to me. Did you try my pipeline in your environment? It’s not that I need 32k tasks, it’s the JSON request that is sent to the agent when job runs that seems to max out at 32k length. 

Please try importing attached pipeline which should hopefully replicate the problem. Let me know if it doesn’t.

**Comment by damccorm** [comment]
Hey - I was not able to repro. I couldn't get an exact copy/import because it was reliant on other pipelines in the project for an artifact value I think? Regardless, it didn't allow me to do that, but I created a similarly large pipeline that ran without issue - https://dev.azure.com/damccorm-repros/big%20json/_releaseProgress?_a=release-pipeline-progress&releaseId=1

From the error, it looks like something may be broken with your artifacts and that might be the source of the problem. Any chance you can create a public project and try to repro there?

**Comment by damccorm** [comment]
Closing due to inactivity. Please comment if you'd still like help on this

**Comment by JScott2002** [comment]
We are seeing a similar issue when running certain pipelines with our self-hosted Linux agent.  We have found that by splitting the pipeline into 2 jobs each using its own agent that the issue resolves but this is not ideal and creates more complexity into the pipeline. We also noticed that by removing Dynatrace from the AKS cluster hosting the self hosted agent it appears to resolve as well, but we aren't sure if this is related to Dynatrace itself or related to the load on the cluster.
